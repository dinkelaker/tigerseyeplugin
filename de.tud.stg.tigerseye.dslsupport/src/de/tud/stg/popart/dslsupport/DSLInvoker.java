/**
 * Copyright 2008, Darmstadt University of Technology
 * GNU GENERAL PUBLIC LICENSE version 2.0
 * @author Tom Dinkelaker
 **/
package de.tud.stg.popart.dslsupport;

import groovy.lang.Closure;
import groovy.lang.MissingMethodException;
import groovy.lang.MissingPropertyException;

import java.util.List;
import java.util.Set;

import org.codehaus.groovy.runtime.InvokerHelper;

public class DSLInvoker {

	private final static boolean DEBUG = false; 
//	================================
//	Methods below are no longer used
//	================================
//	/**
//	 * Tries to invoke the first found keyword method in DSL definition list.
//	 */
//	public static Object invokeMethodOnDslDefs(List dslDefinitions, String name, Object args) {
//		if (DEBUG) println "DSLInvoker.invokeMethodOnDslDefs(): \t try to invoke "+name+" "+args+" on dslDefs";
//		Object result = null;
//		boolean found = false;
//		
//		Object dslDefinition = null;
//
//		Iterator it = dslDefinitions.iterator();
//        while (it.hasNext()) {
//			dslDefinition = it.next();
//			if (DEBUG) println "DSLInvoker.invokeMethodOnDslDefs(): \t try to invoke "+name+" "+args+" on "+dslDefinition;
//		    try {
//		    	//result = tryInvokeMethodOnDslDefinition(dslDefinition, name, args);
//		    	result = InvokerHelper.invokeMethod(dslDefinition,name, args);
//		    	found = true;
//		    } catch (MissingMethodException e1) {
//		    	//do nothing try next dsl
//		    }
//	    	if (found) break;
//		}
//		if (!found) {
//		    //if (DEBUG) println "DSLInvoker.invokeMethodOnDslDefs(): \t not found "+name+" "+args+" in dslDefs.";
//			//throw new MissingMethodException(name,DSLInvoker.class,args);
//			throw new RuntimeException(); //TODO Check exception
//		} else {
//  	        if (DEBUG) println "DSLInvoker.invokeMethodOnDslDefs(): \t found "+name+" "+args+" in "+dslDefinition.class;
//		    return result;
//		}
//	}
//
//	/**
//	 * Tries to invoke the keyword method on a DSL definition.
//	 */
//	private static Object tryInvokeMethodOnDslDefinition(Object dslDefinition, String name, Object args) {
//	    if (DEBUG) println "DSLInvoker.tryInvokeMethodOnDslDefinition(): \t try to invoke on "+name+" "+dslDefinition.class;
//	    try {
//	    	Object result = InvokerHelper.invokeMethod(dslDefinition,name, args);
//		    if (DEBUG) println "DSLInvoker.tryInvokeMethodOnDslDefinition(): \t found "+name+"="+result+" in "+dslDefinition.class;
//	        return result
//		} catch (MissingMethodException e1) {
//		    if (DEBUG) println "DSLInvoker.tryInvokeMethodOnDslDefinition(): \t not found "+name+" "+args+" on "+dslDefinition.class;
//			//do nothing an try next dsl definition
//		}
//	}
//	
//	/**
//	 * Tries to get the first found keyword literal in DSL definition list.
//	 */
//	public static Object getPropertyOnDslDefs(List dslDefinitions, String _property) {
//		Object result = null;
//		boolean found = false;
//		
//		Object dslDefinition = null;
//
//		Iterator it = dslDefinitions.iterator();
//		if (DEBUG && (!it.hasNext())) println "DSLInvoker: empty DSL list";
//        while (it.hasNext()) {
//			dslDefinition = it.next();
//			if (DEBUG) println "DSLInvoker.getPropertyOnDslDefs(): \t try to invoke "+_property+" on "+dslDefinition.class;
//		    try {
//		    	//result = tryGetPropertyOnDslDefinition(dslDefinition, _property);
//		    	result = InvokerHelper.getProperty(dslDefinition,_property);
//		    	found = true;
//		    } catch (MissingPropertyException e1) {
//				//do nothing an try next dsl definition
//		    }
//	    	if (found) break;
//		}
//		if (!found) {
//		    //if (DEBUG) println "DSLInvoker.invokeMethodOnDslDefs(): \t not found "+_property+" in dslDefs.";
//			//throw new MissingPropertyException(_property,DSLInvoker.class);
//			throw new RuntimeException();
//		} else {
//  	        if (DEBUG) println "DSLInvoker.getPropertyOnDslDefs(): \t found "+_property+" in "+dslDefinition.class;
//		    return result;
//		}
//	}
//		
//	/**
//	 * Tries to get the keyword literal in DSL definition.
//	 */
//	private static Object tryGetPropertyOnDslDefinition(Object dslDefinition, String _property){
//	    if (DEBUG) println "DSLInvoker.tryGetPropertyOnDslDefinition(): \t try getting "+_property+" from "+dslDefinition.class;
//		
//		try {
//			Object result = InvokerHelper.getProperty(dslDefinition,_property);
//		    if (DEBUG) println "DSLInvoker.tryGetPropertyOnDslDefinition(): \t found "+_property+"="+result+" in "+dslDefinition.class;
//		    return result;
//		} catch (MissingPropertyException e00) {
//			//do nothing an try next dsl definition
//		}
//	}
//	================================
//	Methods above are no longer used
//	================================
	
	/**
	 * Zips the result of all keyword methods in DSL definition list.
	 */
	public static Object zipMethodOnDslDefs(Set<DSL> dslDefinitions, Closure zipWithClosure, String name, Object args) {
		if (DEBUG) System.out.println("DSLInvoker.zipMethodOnDslDefs(): \t zips method results of "+name+" "+args+" on dslDefs");
		assert zipWithClosure.getMaximumNumberOfParameters() == dslDefinitions.size();

		List results = new java.util.LinkedList(); 
		

		for(Object dslDefinition : dslDefinitions){
			if (DEBUG) System.out.println("DSLInvoker.zipMethodOnDslDefs(): \t try to invoke "+name+" "+args+" on "+dslDefinition);
		    try {
		    	Object result = InvokerHelper.invokeMethod(dslDefinition,name, args);
		    	results.add(result);
		    } catch (MissingMethodException e1) {
		    	results.add(null);
		    }
		}
		
		if (DEBUG) System.out.println("DSLInvoker.zipMethodOnDslDefs(): \t results=$results");
        Object result = zipWithClosure.call(results.toArray());	
        if (DEBUG) System.out.println("DSLInvoker.zipMethodOnDslDefs(): \t final result=$result");
        return result;
	}
	
	/**
	 * Zips the value of all keyword literal in DSL definition list.
	 */
	public static Object zipPropertyOnDslDefs(Set<DSL> dslDefinitions, Closure zipWithClosure, String _property) {
		if (DEBUG) System.out.println("DSLInvoker.zipMethodOnDslDefs(): \t zips property values of "+_property+" on dslDefs");
		assert zipWithClosure.getMaximumNumberOfParameters() == dslDefinitions.size();
		
		List results = new java.util.LinkedList();
		
		for(Object dslDefinition : dslDefinitions){
			if (DEBUG) System.out.println("DSLInvoker.zipPropertyOnDslDefs(): \t try to invoke "+_property+" on "+dslDefinition.getClass());
		    try {
		    	Object result = InvokerHelper.getProperty(dslDefinition,_property);
		    	results.add(result);
		    } catch (MissingPropertyException e1) {
				results.add(null);
		    }
		}
  	    if (DEBUG) System.out.println("DSLInvoker.zipPropertyOnDslDefs(): \t results="+results);
  	    Object result = zipWithClosure.call(results.toArray());
  	  	if (DEBUG) System.out.println("DSLInvoker.zipPropertyOnDslDefs(): \t final result="+result);
		return result;
	}
		
}
