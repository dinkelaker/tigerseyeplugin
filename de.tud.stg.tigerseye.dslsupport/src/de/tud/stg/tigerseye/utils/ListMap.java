package de.tud.stg.tigerseye.utils;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * Supports a <i>map</i> style list transformation as known from functional
 * languages. The transformation functionality is defined in the passed
 * {@link Transformer}.
 * 
 * @author Leo Roos
 * 
 */
public class ListMap {

	public static <A, B> List<B> map(Collection<A> arrayList,
			Transformer<A, B> t) {
		ArrayList<B> result = new ArrayList<B>(arrayList.size());
		for (A a : arrayList) {
			result.add(t.transform(a));
		}
		return result;
	}

}
