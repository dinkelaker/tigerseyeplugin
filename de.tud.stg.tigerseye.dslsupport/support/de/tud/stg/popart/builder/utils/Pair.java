package de.tud.stg.popart.builder.utils;
import java.io.Serializable;

import de.tud.stg.tigerseye.dslsupport.logger.DSLSupportLogger;

/**
 * {@link Pair} links an object of type X and an object of type Y to a pair.
 * 
 * @param <X>
 * @param <Y>
 */

public class Pair<X, Y> implements Serializable {

	private static final long serialVersionUID = -1343477089401295365L;

	private final X x;
	private final Y y;

	/**
	 * @param x
	 * @param y
	 */
	public Pair(X x, Y y) {
		this.x = x;
		this.y = y;
	}

	/**
	 * @return the x
	 */
	public X getX() {
		return this.x;
	}

	/**
	 * @return the y
	 */
	public Y getY() {
		return this.y;
	}

	@Override
	public String toString() {
		return "(" + this.getX() + ", " + this.getY() + ")";
	}

	@Override
	public boolean equals(Object obj) {
		if (obj instanceof Pair<?, ?>) {
			Pair<?, ?> p = (Pair<?, ?>) obj;
			return (p.x.equals(this.x) && (p.y.equals(this.y)));
		} else {
			return false;
		}
	}

	@Override
	public int hashCode() {
		int hash = 7;
		hash = 31 * hash + this.x.hashCode();
		hash = 31 * hash + this.y.hashCode();
		return hash;
	}

	// public static void main(String[] args) {
	// Pair<Integer, Integer> p = new Pair<Integer, Integer>(4, 4);
	// Pair<Integer, Integer> q = new Pair<Integer, Integer>(4, 4);
	// logger.info(p.hashCode());
	// logger.info(q.hashCode());
	// }
}