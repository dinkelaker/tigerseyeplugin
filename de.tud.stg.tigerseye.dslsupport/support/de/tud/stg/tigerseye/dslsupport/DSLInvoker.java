package de.tud.stg.tigerseye.dslsupport;

import groovy.lang.Closure;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

import de.tud.stg.tigerseye.dslsupport.logger.DSLSupportLogger;

public final class DSLInvoker{
	
	private static final String NO_CONTEXT = "no_context_information";

	private static final DSLSupportLogger logger = new DSLSupportLogger(DSLInvoker.class);
	
	// TODO For Issue #37 (Leo_Roos;Nov 18, 2011)
	// In the current version, the DSL instances are shared between different DSL files.
	// Shouldn't cache instances, they might change between files, need additional information about the file _and_ class to have an exact instance definition
	private static final ConcurrentHashMap<dslObjectKey<?>, DSL> invokers = new ConcurrentHashMap<dslObjectKey<?>, DSL>();

	public static Object eval(Class<? extends DSL> clazz, Closure<?> c) {
		return eval(NO_CONTEXT,toList(clazz),c);
	}
	
	public static Object eval(String contextInformation, Class<? extends DSL> clazz, Closure<?> c) {
		return eval(contextInformation, toList(clazz), c);
	}

	public static Object eval(Class<? extends DSL>[] clazzes, Closure<?> c) {
		return eval(NO_CONTEXT,Arrays.asList(clazzes) , c);		
	}
	
	public static Object eval(List<Class<? extends DSL>> clazzes, Closure<?> c) {
		return eval(NO_CONTEXT, clazzes, c);
	}
	
	public static Object eval(String contextInformation, List<Class<? extends DSL>> clazzes, Closure<?> c) {
		
		Set<DSL> dsls = new HashSet<DSL>();

		for (Class<? extends DSL> clazz : clazzes) {
			DSL dsl = getDSL(clazz);
			if (dsl != null)
				dsls.add(dsl);
		}

		InterpreterCombiner ic = new InterpreterCombiner(dsls, new HashMap<String, Object>()); //TODO For Issue #36, sort the DSL instances "dsls"
		return ic.eval(c);
	}
	
	public static <T extends DSL> T getDSL(Class<T> clazz) {
		return getDSL(NO_CONTEXT, clazz);
	}

	public static <T extends DSL> T getDSL(String contextInformation, Class<T> clazz) {
		return getDSL(new dslObjectKey<T>(contextInformation, clazz));
	}

	private static List<Class<? extends DSL>> toList(Class<?>... clazzes) {
		List<Class<? extends DSL>> result = new ArrayList<Class<? extends DSL>>(clazzes.length);
		for (Class<?> c : clazzes) {
			/* Throws exception if not of type DSL */
			Class<? extends DSL> clazz = (Class<? extends DSL>) c;
			result.add(clazz);
		}
		return result;
	}

	private static <T extends DSL> T getDSL(dslObjectKey<T> invokerKey) {				
		DSL object = invokers.get(invokerKey);

		if (object == null) {
			try {
				object = invokerKey.clazz.newInstance();
			} catch (InstantiationException e) {
				logger.error("Failed to instantiate class", e);
			} catch (IllegalAccessException e) {
				logger.error("Failed to instantiate class", e);
			}
			
			invokers.put(invokerKey, object);
		}
		
		if(object == null){
			throw new IllegalArgumentException("No instance for "+ invokerKey + " was found, nor could it be instantiated.");
		}
		
		return (T) object;
	}

	static DSL getInvoker(Class<? extends DSL> clazz) {
		return getDSL(clazz);
	}
	
	private static class dslObjectKey<T>{
		
		private final String context;
		private final Class<T> clazz;

		public dslObjectKey(String context, Class<T> clazz) {
			this.context = context;
			this.clazz = clazz;
		}
		
		@Override
		public boolean equals(Object obj) {
		    	if (obj == null){
		    		return false;
		    	}
		    	if (obj == this){
		    		return true;
		    	}
		    	if (obj.getClass() != this.getClass()) {
		    		return false;
		    	}
		    	DSLInvoker.dslObjectKey<?> other = (DSLInvoker.dslObjectKey<?>) obj;
		    	boolean equal = context == other.context && clazz.equals(other.clazz); 
		    	return equal;
		    }
		
		@Override
		public int hashCode() {
			return context.hashCode() * 11 + 7  * clazz.hashCode();
		}
		
		@Override
		public String toString() {
			return "DSLObjectKey[context="+context+",class="+clazz+"]";
		}
	}
	
/*
	private Class<? extends DSL> clazz;

	private DSLInvoker() {
		//Not supposed to be instantiated
	}

	/**
	 * @deprecated use static methods for consistent access
	 //
	@Deprecated
	public DSLInvoker(Class<? extends DSL> clazz) {
		this.clazz = clazz;
	}

	/**
	 * @deprecated use static methods for consistent access
	 
	@Deprecated
	public Object eval(Closure<?> c) {
		if (this.clazz == null)
			throw new IllegalStateException("If instance eval is used the Invoker must have been initialized with a clazz");
		else
			return eval(this.clazz, c);
	}
	*/
	
}
