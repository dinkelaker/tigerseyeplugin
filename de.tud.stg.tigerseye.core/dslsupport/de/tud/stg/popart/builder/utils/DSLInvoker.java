package de.tud.stg.popart.builder.utils;

import groovy.lang.Closure;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import de.tud.stg.popart.dslsupport.DSL;
import de.tud.stg.popart.dslsupport.InterpreterCombiner;

public class DSLInvoker {
	private static final DSLInvoker instance = new DSLInvoker();

	private static final Map<Class<? extends DSL>, DSL> invokers = new HashMap<Class<? extends DSL>, DSL>();

	private final Class<? extends DSL> clazz;

	public static <T extends DSL> T getDSL(Class<T> clazz) {

		DSL object = invokers.get(clazz);

		if (object == null) {
			try {
				object = clazz.newInstance();
			} catch (InstantiationException e) {
				e.printStackTrace();
			} catch (IllegalAccessException e) {
				e.printStackTrace();
			}
			invokers.put(clazz, object);
		}

		return (T) object;

	}

	private DSLInvoker() {
		this.clazz = null;
	}

	public DSLInvoker(Class<? extends DSL> clazz) {
		this.clazz = clazz;
	}

	public static DSLInvoker getDSLInvoker() {
		return instance;
	}

	public static DSL getInvoker(Class<? extends DSL> clazz) {
		return getDSL(clazz);
	}

	public Object eval(Closure c) {
		if (this.clazz != null)
			return eval(this.clazz, c);
		else
			throw new IllegalStateException();
	}

	public static Object eval(Class<? extends DSL> clazz, Closure c) {
		DSL dsl = getDSL(clazz);

		c.setDelegate(dsl);
		c.setResolveStrategy(Closure.DELEGATE_FIRST);
		return c.call();
	}

	public static Object eval(Class<? extends DSL>[] clazzes, Closure c) {
		List<DSL> dsls = new LinkedList<DSL>();

		for (Class<? extends DSL> clazz : clazzes) {
			dsls.add(getDSL(clazz));
		}

		InterpreterCombiner ic = new InterpreterCombiner(dsls, new HashMap());
		return ic.eval(c);
	}

	public static <T> T[] asArray(T... ts) {
		return ts;
	}

	public static <T> List<T> asList(T... ts) {
		ArrayList<T> list = new ArrayList<T>(ts.length);
		for (T t : ts) {
			list.add(t);
		}
		return list;
	}

	public static void main(String[] args) {
		List<Integer>[] asArray = asArray(new ArrayList<Integer>(), new ArrayList<Integer>(), new ArrayList<Integer>());

		List<ArrayList<Integer>> asList = Arrays.asList(new ArrayList<Integer>(), new ArrayList<Integer>(),
				new ArrayList<Integer>());

	}
}
