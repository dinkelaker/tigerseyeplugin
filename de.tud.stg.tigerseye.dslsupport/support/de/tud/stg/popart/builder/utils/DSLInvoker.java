package de.tud.stg.popart.builder.utils;

import groovy.lang.Closure;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import de.tud.stg.popart.dslsupport.DSL;
import de.tud.stg.popart.dslsupport.InterpreterCombiner;

public class DSLInvoker {
	
	private static final Logger logger = LoggerFactory
			.getLogger(DSLInvoker.class);
	
	private static final DSLInvoker instance = new DSLInvoker();

	private static final Map<Class<? extends DSL>, DSL> invokers = new HashMap<Class<? extends DSL>, DSL>();

	public static <T> T[] asArray(T... ts) {
		return ts;
	}

	public static <T> List<T> asList(T... ts) {
		ArrayList<T> list = new ArrayList<T>(ts.length);
		for (T t : ts) {
			list.add(t);
		}
		return list;
	}

	public static Object eval(Class<? extends DSL> clazz, Closure c) {
		DSL dsl = getDSL(clazz);

		c.setDelegate(dsl);
		c.setResolveStrategy(Closure.DELEGATE_FIRST);
		return c.call();
	}

	public static Object eval(Class<? extends DSL>[] clazzes, Closure c) {
		List<DSL> dsls = new LinkedList<DSL>();

		for (Class<? extends DSL> clazz : clazzes) {
			dsls.add(getDSL(clazz));
		}

	/*
	 * FIXME(Leo Roos): Must use the @deprecated method until the runtime
	 * jars, which also package a version of an InterpreterCombiner, are
	 * generated from the same source this class accesses during
	 * development.
	 */
	InterpreterCombiner ic = new InterpreterCombiner(dsls,
		new HashMap<String, Object>());
		return ic.eval(c);
	}

	public static <T extends DSL> T getDSL(Class<T> clazz) {

		DSL object = invokers.get(clazz);

		if (object == null) {
			try {
				object = clazz.newInstance();
			} catch (InstantiationException e) {
				logger.error("Failed to instantiate class", e);
			} catch (IllegalAccessException e) {
				logger.error("Failed to instantiate class", e);
			}
			invokers.put(clazz, object);
		}

		return (T) object;

	}

	public static DSLInvoker getDSLInvoker() {
		return instance;
	}

	public static DSL getInvoker(Class<? extends DSL> clazz) {
		return getDSL(clazz);
	}

	private final Class<? extends DSL> clazz;

	private DSLInvoker() {
		this.clazz = null;
	}

	public DSLInvoker(Class<? extends DSL> clazz) {
		this.clazz = clazz;
	}

	public Object eval(Closure c) {
		if (this.clazz != null)
			return eval(this.clazz, c);
		else
			throw new IllegalStateException("Closure must not be null");
	}

}
