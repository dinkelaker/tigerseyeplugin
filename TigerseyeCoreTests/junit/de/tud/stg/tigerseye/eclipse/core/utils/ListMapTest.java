package de.tud.stg.tigerseye.eclipse.core.utils;

import static org.junit.Assert.assertThat;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import de.tud.stg.tigerseye.utils.LMTransformer;
import de.tud.stg.tigerseye.utils.ListBuilder;
import de.tud.stg.tigerseye.utils.ListMap;
import static junit.framework.Assert.*;

public class ListMapTest {

	@Mock
	LMTransformer<String, Integer> sToI;

	@Before
	public void setUp() {
		MockitoAnnotations.initMocks(this);
	}

	@Test
	public void testTransformEmptyList() throws Exception {
		List<Integer> result = ListMap.map(new ArrayList<String>(), sToI);
		assertEquals(0,result.size());
	}

	@Test
	public void testTransformertransforms() throws Exception {

		LMTransformer<Integer, String> t = new LMTransformer<Integer, String>() {

			@Override
			public String transform(Integer input) {
				return input.toString();
			}

		};

		String transform = t.transform(2);
		assertEquals("2",transform);

	}

	@Test
	public void testTransformSomeStringsToInteger() throws Exception {
		ArrayList<String> arrayList = new ArrayList<String>();
		Collections.addAll(arrayList, "1", "2", "3", "4");
		List<Integer> result = ListMap.map(arrayList,
				new LMTransformer<String, Integer>() {
					@Override
					public Integer transform(String input) {
						return Integer.parseInt(input);
					}
				});
		assertTrue(result.containsAll(ListBuilder.begin(1).add(2).add(3).add(4).toList()));
	}

}
