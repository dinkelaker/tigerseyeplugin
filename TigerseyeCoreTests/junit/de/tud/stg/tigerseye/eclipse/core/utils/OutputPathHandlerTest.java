package de.tud.stg.tigerseye.eclipse.core.utils;

import static org.junit.Assert.assertEquals;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import java.io.File;

import org.eclipse.core.resources.IFile;
import org.eclipse.core.resources.IProject;
import org.eclipse.core.runtime.IPath;
import org.eclipse.core.runtime.Path;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import utilities.TestUtilities;
import de.tud.stg.tigerseye.eclipse.core.builder.transformers.FileType;
import de.tud.stg.tigerseye.eclipse.core.utils.OutputPathHandler;

public class OutputPathHandlerTest {
    private static final String OUTPUT_DIRECTORY = "src-testoutput";

    static {
	TestUtilities.initLogger();
    }
    
    @Before
    public void before(){
	actualOutputPath = null;
    }

    private static final Logger logger = LoggerFactory
	    .getLogger(OutputPathHandlerTest.class);

    @Test
    public void testGetFileOutputPathSetDsl() {
	File originalSrcProjRelPath = new File(
		"some/package/SimpleSet.set.dsl");
    	final File expectedOutputPat = new File(
		"some/package/"
			+ "SimpleSet$_set_dsl.groovy");
	testCorrectConversion(originalSrcProjRelPath, expectedOutputPat);
    }

    @Test
    public void testGetFileOutputPathGroovyfile() {
	File originalSrcProjRelPath = new File(
		"some/package/SimpleSet.groovy.dsl");
	final File expectedOutputPat = new File(
		"some/package/"
			+ "SimpleSet.groovy");
	testCorrectConversion(originalSrcProjRelPath, expectedOutputPat);
    }

    @Test
    public void testGetFileOutputPathJavaFile() {
	File originalSrcProjRelPath = new File("SimpleSet."+FileType.JAVA.srcFileEnding);
	final File expectedOutputPat = new File(
		"SimpleSet." + FileType.JAVA.outputFileEnding);
	testCorrectConversion(originalSrcProjRelPath, expectedOutputPat);
    }

    private void testCorrectConversion(File originalSrcRelPath,
	    final File expectedOutputPath) {
	
	Path originialPath = new Path(originalSrcRelPath.getPath());	
	IPath fileOutputPath = new OutputPathHandler(OUTPUT_DIRECTORY).getSrcRelativeOutputPath(originialPath);
	File generatedOutputFile = fileOutputPath.toFile();
	logger.debug("for path {} was generated {} ", originialPath, fileOutputPath);
	assertEquals("Generated name does not equal expected",
		expectedOutputPath, generatedOutputFile);
    }
    
    private static IPath actualOutputPath = null;
    
    private IFile getFileMock(IPath path, IProject project){
    	IFile resource = mock(IFile.class);
    	when(resource.getProjectRelativePath()).thenReturn(path);
    	when(resource.getProject()).thenReturn(project);
    	return resource;
    }
    
    @Test
    public void testOutputFile() throws Exception {
	//setup
	File origin = new File("/src/only/src/matters.sql."+FileType.TIGERSEYE.srcFileEnding);
		
	final IProject project = mock(IProject.class);
	when(project.getFile(any(IPath.class))).thenAnswer(new Answer<IFile>() {

	    @Override
	    public IFile answer(InvocationOnMock invocation) throws Throwable {
		actualOutputPath = (IPath) invocation.getArguments()[0];
		return getFileMock(actualOutputPath, project);
	    }
	});	
	IFile resource = getFileMock(new Path(origin.getPath()), project);
	OutputPathHandler resourceHandler = new OutputPathHandler(OUTPUT_DIRECTORY);
//	when(resourceHandler.getProjectRelativePath(any(IPath.class))).thenReturn(outputDirectory.append(new Path("/forced/Return")));
	//execute
	resourceHandler.getOutputFile(resource);
	//validate
	logger.debug("For origin {} generated {}", origin, actualOutputPath.toFile());
	assertEquals(new File(OUTPUT_DIRECTORY), getRootFolder(actualOutputPath.toFile()));
    }

    private static File getRootFolder(File path) {
	File rootFolder = path;
	File parentFile = path.getParentFile();
	while(parentFile != null){
	    rootFolder = parentFile;
	    parentFile = rootFolder.getParentFile();
	}
	return rootFolder;
    }
    
    public static void main(String[] args) {
	File rootFolder = getRootFolder(new Path("src/only/src/matters.any.dsl").toFile());
	System.out.println(rootFolder);
	
	File rootFolder2 = getRootFolder(new File("other/path/without/"));
	System.out.println(rootFolder2);
    }
    
    @Test
    public void testProjRelativeOutputPath() throws Exception {
	//setup
	String outputDir = "src-tigerseye";
	String origin = new File("/only/src/matters.any.dsl").getPath();
	String expected = new File("/only/src/matters$_any_dsl.groovy").getPath();
	
	Path projRelOriginPath = new Path(origin);	
	OutputPathHandler resourceHandler = new OutputPathHandler(OUTPUT_DIRECTORY);
	resourceHandler.setLocalOutputDirectoryName(outputDir);
	//execute
	IPath actual = resourceHandler.getSrcRelativeOutputPath(projRelOriginPath);
	//validate
	logger.debug("For origin {} generated {}", origin, actual.toFile());
	assertEquals(new File(expected), actual.toFile());
    }
    
    @Test
	public void testGetSrcNameForOutputNameJava() throws Exception {		    
    	String outputName = "myjavafile.java";
    	String expected = "myjavafile.java.dsl";    	    	
		assertCorrectOutputforSourceName( outputName, expected);
	}
    
    @Test
	public void testGetSrcNameForOutputNamePureGroovy() throws Exception {		    
    	String outputName = "myjavafile.groovy";
    	String expected = "myjavafile.groovy.dsl";    	    	
		assertCorrectOutputforSourceName(outputName, expected);
	}
    
    @Test
	public void testGetSrcNameForOutputNameDsl1() throws Exception {		    
    	String outputName = "myjavafile$_sql_dsl.groovy";
    	String expected = "myjavafile.sql.dsl";    	    	
		assertCorrectOutputforSourceName(outputName, expected);
	}
    
    @Test
	public void testGetSrcNameForOutputNameDsl2() throws Exception {		    
    	String outputName = "myjavafile$_sql_set_cond_foreach_dsl.groovy";
    	String expected = "myjavafile.sql.set.cond.foreach.dsl";    	    	
		assertCorrectOutputforSourceName(outputName, expected);
	}
    
    @Test
	public void testGetSrcNameForOutputNameIllegaluseofdsl() throws Exception {		    
    	String outputName = "myjavafile$_sql_set_cond_foreach_dsl.groovy.notexistent";    	    	
		assertCorrectOutputforSourceName(outputName, null);
	}
    
    @Test
	public void testGetSrcNameForOutputNameIllegalusedslformat() throws Exception {		    
    	String outputName = "myjavafile$_sql_set_cond_foreach.dsl.groovy";	    	
		assertCorrectOutputforSourceName( outputName, null);
	}
    
    @Ignore("Not sure if this should be treated as illegal format")
    @Test(expected=IllegalArgumentException.class)
	public void testGetSrcNameForOutputNameIllegalformat2() throws Exception {		    
    	String outputName = "myjavafile$sql_set_cond_foreach_dsl.groovy";	    	
		assertCorrectOutputforSourceName(outputName, "");
	}

	private void assertCorrectOutputforSourceName(String outputName, String expected) {
		OutputPathHandler outputPathHandler = new OutputPathHandler(OUTPUT_DIRECTORY);
    	String actual = outputPathHandler.getSourceNameForOutputName(outputName);
    	assertEquals(expected, actual);
	}

    
}
