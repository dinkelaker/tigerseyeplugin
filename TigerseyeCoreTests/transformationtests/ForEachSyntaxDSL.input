package generated;

import de.tud.stg.popart.builder.utils.DSLInvoker;
import de.tud.stg.popart.builder.test.dsls.ForEachSyntaxDSL;
	 
new DSLInvoker(ForEachSyntaxDSL.class).eval() {
	String[] abc = ["a", "b", "ac"];
	
	for ( String s : abc ) {
		if (s.startsWith("a")) {
			System.out.println(s);
		}
	}
}