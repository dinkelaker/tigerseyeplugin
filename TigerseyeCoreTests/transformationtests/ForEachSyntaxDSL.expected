package generated;

import de.tud.stg.popart.builder.utils.DSLInvoker;
import de.tud.stg.popart.builder.test.dsls.ForEachSyntaxDSL;
	 
new DSLInvoker(ForEachSyntaxDSL.class).eval() {

	String[] abc = ["a", "b", "ac"];
	
	forEach(
			String.class,
			"s",
			abc,
			{
				if (s.startsWith("a")) {
					System.out.println(s);
				}
			})
}